### PR TITLE
Subtypes of unions of functions

### PR DESCRIPTION
**Pull request by mwalkerwells**
Adding a new example to the "Subtypes of complex types" section. Hopefully people find this helpful!

### CODE REVIEW

File: website/en/docs/lang/subtypes.md [hunk 1]
```diff
@@ -136,3 +136,37 @@ type Func2 = (number, string) => void;
 let func1: Func1 = (a: number) => {};
 let func2: Func2 = func1;
 ```
+
+#### Subtypes of unions of functions <a class="toc" id="toc-subtypes-of-unions-of-functions" href="#toc-subtypes-of-unions-of-functions"></a>
+
+A function is a subtype of another when the input is contravariant (_less specific type / supertype_) and the output is covariant (_more specific type / subtype_).
```
Comment (samwgoldman) [inline website/en/docs/lang/subtypes.md]: The wording here makes me imagine that some functions don't have contravariant inputs and covariant outputs, but all functions have this behavior.

It's more precise to say that function types are contravariant with respect to their parameter types and covariant with respect to their return type. That is, contra- and covariance are properties of the function type.

However, I think what you have is easier to understand. I would just remove the variance stuff (at least to start with):

"A function is a subtype of another when its argument types are less specific (supertype) or its return type is more specific (subtype)."

You can then go on to introduce the technical jargon: "In technical terms, we say that function types are contravariant with respect to their parameter types and covariant with respect to their return type."

File: website/en/docs/lang/subtypes.md [hunk 2]
```diff
@@ -136,3 +136,37 @@ type Func2 = (number, string) => void;
...
+
+```js
+// @flow
+declare var foo1: (string => ?string) => void
+declare var bar1: ?string => string
+
+foo1(bar1) // Works!
```
Comment (samwgoldman) [inline website/en/docs/lang/subtypes.md]: This example gets straight to the point, but I think a concrete example would be better at building intuition.

A lot of javascript developers are familiar with callback functions, so how about examples using that?

```
function f(callback: (name: string) => void) {
  callback("Flow");
}

function g(name: ?string): void {
  let name = name || "World";
  console.log("Hello, " + name);
}

f(g); // `?string => void` is a subtype of `string => void`
```

```
function f(callback: () => ?string) {
  var name = callback() || "World";
  console.log("Hello, " + name);
}

function g(): string {
  return "Flow";
}

f(g); // `() => string` is a subtype of `() => ?string`
```

File: website/en/docs/lang/subtypes.md [hunk 3]
```diff
@@ -136,3 +136,37 @@ type Func2 = (number, string) => void;
...
+```
+So what would be a subtype of a union of functions? Well we know that a subtype of ``A | B`` woudld be either ``A`` or ``B``, but we are looking for the subtype of ``A => B | C => D``. Let's walk through this...
```
Comment (samwgoldman) [inline website/en/docs/lang/subtypes.md]: To be honest, I think this is a fun exercise for a blog post, but I'm worried that this example is too abstract and might discourage folks from grasping the concepts if introduced too early.

File: website/en/docs/lang/subtypes.md [hunk 4]
```diff
@@ -136,3 +136,37 @@ type Func2 = (number, string) => void;
...
+foo2(baz2); // Works!
+```
+But is there anything else we can pass into ``foo2``? Yes! 
+```js
+// @flow
+
+declare var bar3: (A | C) => (B & D)
+foo2(bar3);
+```
+A subtype of a union of functions is a union (``|``) of the inputs and a intersection (``&``) of the outputs. While not covered here, the inverse is true for a intersection of functions: a subtype of a intersection of functions (``A => B & C => D``) is a intersection (``&``) of the inputs and a union (``|``) of the outputs.
```
Comment (samwgoldman) [inline website/en/docs/lang/subtypes.md]: This *is* a very cool derivation, though!

**Comment by mwalkerwells** [comment]
> I'm not sure about including the (A|C => B&D) <: (A => B) | (C => D) example in the docs, because I'm not sure how helpful it is for most people. Very cool stuff, though!

Yeah, it definitely would not have been helpful to me when I was first learning, but it was a really helpful realization when I was writing a library that heavily relied on passing functions as values & expanding the type via ```(...rest: Array<Fn>)```.

I'd love to see an "Advanced" section that this sort of thing could be a part of. For now, I'll drop it. :)

File: website/en/docs/lang/subtypes.md (hunk 2 above)
Comment (mwalkerwells) [inline website/en/docs/lang/subtypes.md]: Yeah callbacks are pretty familiar, but they generally lack any return type. :( It makes it harder to illustrate that subtypes are depended on both input and output.

File: website/en/docs/lang/subtypes.md (hunk 4 above)
Comment (mwalkerwells) [inline website/en/docs/lang/subtypes.md]: Thanks!

**Pull request closed without merging**